src/controller/timeouts.rs

```diff
@@ -0,0 +1,95 @@
+use std::{
+    sync::atomic::{AtomicUsize, Ordering},
+    time::Duration,
+};
+
+use parking_lot::Mutex;
+use tokio::sync::watch;
+
+/// Keeps track of sliding percentile of request durations.
+/// Calculates a timeout so that (1-q) fraction of the slowest requests are retried.
+/// Running and timed out requests count as "infinite" duration.
+pub struct TimeoutManager {
+    quantile: f32,
+    durations: Mutex<Vec<Duration>>,
+    num_infs: AtomicUsize,
+    current_timeout: watch::Sender<Option<Duration>>,
+}
+
+impl TimeoutManager {
+    pub fn new(quantile: f32) -> Self {
+        let (tx, _) = watch::channel(None);
+        Self {
+            quantile,
+            durations: Mutex::new(Vec::new()),
+            num_infs: AtomicUsize::new(0),
+            current_timeout: tx,
+        }
+    }
+
+    pub fn complete_ok(&self, duration: Duration) {
+        let mut durations = self.durations.lock();
+        durations.push(duration);
+        let num_infs = self.num_infs.load(Ordering::Relaxed);
+
+        // TODO: optimize time complexity
+        let kth = ((durations.len() + num_infs) as f32 * self.quantile).floor() as usize;
```

Comment:
This algorithm ensures that the `timeout_percentile` of the slowest requests are retried on average. Its a generalization of what you proposed (wait until 90% of the requests complete, then retry the rest) but in the streaming manner - when the requests could start at different times. In this case I'm tracking the durations of completed requests and the current number of "ongoing" requests. The goal is to find a timeout which would "cut" the `1-timeout_quantile` fraction of the remaining requests.

I totally agree that this algorithm is hard to comprehend. But I couldn't fast a better one with the same performance and the same good qualities, namely:
- the requests are retried as soon as we know they fall into the given fraction of the slowest ones,
- if less than the `timeout_percentile` requests have finished so far, we will continue waiting for any of them to complete, 
- a request which has been just retried should be also timed out with the current duration

Reply:
Fixed, almost as you suggested, but instead of using the longest existing time I always use the hard-coded timeout until there are 50 observations. Otherwise, some unexpected durations (either too fast or too long) could lead to unpredictable consequences.